libcontainer/user/lookup_windows.go

```diff
@@ -7,6 +7,28 @@ import (
 	"strconv"
 )
 
+// userFromOS converts an os/user.(*User) to local User
+//
+// (This does not include Pass, Shell or Gecos)
+func userFromOS(u *user.User) (User, error) {
+	newUser := User{
+		Name: u.Username,
+		Home: u.HomeDir,
+	}
+	id, err := strconv.Atoi(u.Uid)
```

Comment:
I'm wondering if this even _works_ on Windows. Looking at the code; if user is "current" user, it is returned as-is; https://github.com/golang/go/blob/go1.15.5/src/os/user/lookup.go#L32-L37

In other cases, it would look-up the user through a syscall; https://github.com/golang/go/blob/go1.15.5/src/os/user/lookup_windows.go#L322-L331

And construct a user; https://github.com/golang/go/blob/go1.15.5/src/os/user/lookup_windows.go#L290-L297

However, both `Uid` and `Gid` on Windows are actually a SID (https://en.wikipedia.org/wiki/Security_Identifier), which would looks something like `S-1-5-21-3623811015-3361044348-30300820-1013`, or `S-1-5-21domain-513`

Reply:
Looks like 

- in Moby, `libcontainer/user` is only used in unix files.
- In BuildKit it's used in a non-specific file https://github.com/moby/buildkit/blob/f26452a84763b9e13e7e157904cf28008cd58016/executor/oci/user.go
but BuildKit currently doesn't run on Windows; @TBBle may be able to tell if that file is actually used on Windows
- containerd also uses this package in a non-platform specific file https://github.com/containerd/containerd/blob/62083eeed275db8c154dd87a5f84444e97ddb507/oci/spec_opts.go#L539

Reply:
For BuildKit, `GetUser` (and that whole file you linked) is Linux-container specific, since it populates a [linux/solaris-specific fields](https://github.com/opencontainers/runtime-spec/blob/7413a7f753e1bd9a6a9c6dc7f96f55888cbbd476/specs-go/config.go#L88-L91).

Windows containers populate [a username field](https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L96-L97) instead.

So that file would be used in LCOW mode, I guess, on Windows builds, so it can't be unix-specific. Right now BuildKit has a bunch of stuff we'd need for LCOW mode that's in platform-specific source files anyway, but I'm uninclined to make it that worse. Also, last time I tried, LCOW didn't support non-root UIDs anyway. But that was a long time ago, I was hoping WSL2 would fix that.

Looking slightly outwards, the containerd executor in BuildKit does not take care to only call these APIs with a Linux target platform, so attempting to use `USER` in a Dockerfile or the equivalent in LLB will fail for a Windows container, because we pass it through that `GetUser` code path, and a bunch of other unix-target-specific stuff. (Assuming there's not a higher-level rejection floating around). `USER 5:7` would pass the checks in BuildKit, but end up putting data in the linux/solaris specific fields I mentioned above, and _hopefully_ fail the container startup.

I hadn't noticed that, and apparently never tried to test that, so I'll add it to the BuildKit issue tracking WCOW.

Coming back to this code, based on the above, the whole lookup.go exported API only makes sense on unix host platforms, since (if I am skimming this correctly) it's for looking up users and groups on the _host_, I assume in order to resolve users/groups from outside so that the uid/gid can be passed into the spec.

So I'd suggest just dropping lookup.go and lookup_windows.go, and move those exported functions into lookup_unix.go, like other functions, e.g., `CurrentUser()`.

To address the actual linked file, I don't see any code in BuildKit that can call through to *these* APIs. The linked file is using the other part of the API (from user.go) which is about resolving users from inside the container. Its Unix container-specific (relies on Uid as `int`), but makes sense to execute on Windows for LCOW (as it's resolving from an /etc/passwd and /etc/group from inside the container's mounted filesystem).

I had a quick look at the containerd link you have, and similar to BuildKit, it's calling the user.go APIs that care about the container platform, not the host platform.

Reply:
> Coming back to this code, based on the above, the whole lookup.go exported API only makes sense on unix host platforms, since (if I am skimming this correctly) it's for looking up users and groups on the host, I assume in order to resolve users/groups from outside so that the uid/gid can be passed into the spec.
>
> So I'd suggest just dropping lookup.go and lookup_windows.go, and move those exported functions into lookup_unix.go, like other functions, e.g., CurrentUser().

Thanks for looking! I was thinking in that direction as well; better to explicitly not implement (or alternatively, add a "stub" implementation that produces an error when called) than to have something that gives the impression it's implemented.

Reply:
Yeah, I was thinking about an always-error stub, but since we don't seem to have any current callers in Moby, BuildKit, or containerd, I'm not sure it's worth trying to maintain the API.

But perhaps we do need stubs, _because_ we want to not regress the API, faulty as it may be. Anyone calling this on a Windows platform would have been unavoidably getting an error from `Atoi` anyway.

Reply:
There's been some discussions in this repository about libcontainer not being intended for external consumption (which is why I did some refactoring in other repositories to reduce it's usage); that said, `libcontainer/user` is used in quite some places. containerd is not at v1.0.0 yet, so breaking changes are (technically) still allowed, so from that perspective making them _now_ (before v1.0.0) and rip off the bandaid could be preferable (if libcontainer is always gonna be intended for _linux_ / _unix_ only, and not for Windows).

I'm open to both approaches though (heh, and not a maintainer on this repository 😅)

Reply:
So, do we gravitate towards removing this code? I am all for it! Removed (and a couple other windows things).